packages/collaboration-manager/src/OperationsTransformer.ts

```diff
@@ -0,0 +1,325 @@
+     */
+    if (operation.index.blockIndex === undefined) {
+      return Operation.from(operation);
+    }
+
+    /**
+     * Check that againstOp affects current operation
+     */
+    if (againstOp.index.blockIndex! > operation.index.blockIndex!) {
+      return Operation.from(operation);
+    }
+
+    /**
+     * Update the index of the current operation
+     */
+    switch (againstOp.type) {
+      /**
+       * Cover case 1
+       */
+      case OperationType.Insert:
+        newIndexBuilder.addBlockIndex(operation.index.blockIndex! + 1);
+        break;
+
+      /**
+       * Cover case 2
+       */
+      case OperationType.Delete:
+        if (operation.index.blockIndex !== undefined) {
+          if (againstOp.index.blockIndex! < operation.index.blockIndex!) {
+            newIndexBuilder.addBlockIndex(operation.index.blockIndex! - 1);
+          } else {
+            return new Operation(OperationType.Neutral, newIndexBuilder.build(), { payload: [] }, operation.userId, operation.rev);
+          }
+        }
+
+        break;
+
+      /**
+       * Cover case 3
+       */
+      default:
+        return Operation.from(operation);
+    }
+
+    /**
+     * Return new operation with the updated index
+     */
+    const newOp = Operation.from(operation);
+
+    newOp.index = newIndexBuilder.build();
+
+    return newOp;
+  }
+
+  /**
+   * Method that transforms operation against text operation
+   *
+   * Cases:
+   * 1. Current operation is a block operation
+   *    - Text opearation cant affect block operation so return copy of the current one
+   *
+   * 2. Current operation is a text operation
+   *    - Against operation is Insert
+   *        - Transform current operation against textInsert
+   *    - Against operation is Delete
+   *        - Check that againstOp affects current operation and transform against text operation
+   *    - Against operation is Modify or Neutral
+   *        - Modify and Neutral operations do not affect any of the text operations so return copy of the current operation
+   *
+   * @param operation - Operation to be transformed
+   * @param againstOp - Operation against which the current operation should be transformed
+   * @returns {Operation<OperationType>} new operation
+   */
+  #transformAgainstTextOperation<T extends OperationType>(operation: Operation<T>, againstOp: Operation<OperationType>): Operation<T> | Operation<OperationType.Neutral> {
+    const index = operation.index;
+    const againstIndex = againstOp.index;
+
+    /**
+     * Cover case 1
+     */
+    if (index.isBlockIndex) {
+      return Operation.from(operation);
+    }
+
+    const sameInput = index.dataKey === againstIndex.dataKey;
+    const sameBlock = index.blockIndex === againstIndex.blockIndex;
+
+    /**
+     * Check that againstOp affects current operation
+     */
+    if (sameInput && sameBlock && againstIndex.textRange![0] > index.textRange![1]) {
+      return Operation.from(operation);
+    }
```

Comment:
You can extend condition here to check if it's not the same input or not the same block